src/components/NewTagDialog.tsx

```diff
@@ -0,0 +1,473 @@
+  /**
+   * Callback to store a reference of the rendered <li> element in DialogBoxCommitGraph
+   */
+  setRef: (el: HTMLLIElement) => void;
+
+  /**
+   * Update function for baseCommitId state.
+   */
+  updateBaseCommitId: React.Dispatch<React.SetStateAction<string>>;
+}
+
+export const DialogBoxCommitGraph: React.FunctionComponent<
+  IDialogBoxCommitGraphProps
+> = (props: IDialogBoxCommitGraphProps): React.ReactElement => {
+  const { filter, baseCommitId, updateBaseCommitId } = props;
+  const pastCommits = props.pastCommits
+    .filter(
+      commit =>
+        !filter ||
+        commit.commit_msg.toLowerCase().includes(filter.toLowerCase())
+    )
+    .slice();
+
+  const [nodeHeights, setNodeHeights] = React.useState<{
+    [sha: string]: number;
+  }>({});
+  const nodes = React.useRef<{ [sha: string]: HTMLLIElement }>({});
+
+  React.useEffect(() => {
+    const resizeObserver = new ResizeObserver(entries => {
+      for (const entry of entries) {
+        const borderBoxSize = Array.isArray(entry.borderBoxSize)
+          ? entry.borderBoxSize[0]
+          : entry.borderBoxSize;
+        const offsetHeight = borderBoxSize.blockSize;
+        const sha = entry.target.id;
+        setNodeHeights(prev => ({ ...prev, [sha]: offsetHeight }));
+      }
+    });
+
+    pastCommits.forEach(commit => {
+      const node = nodes.current[commit.commit];
+      resizeObserver.observe(node, {
+        box: 'border-box'
+      });
+    });
+
+    return () => resizeObserver.disconnect();
+  }, [pastCommits]);
+
+  let isFilter;
+  if (filter === '') {
+    isFilter = true;
+  }
+
+  return (
+    <div className={historyDialogBoxWrapperStyle}>
+      {isFilter && (
+        <GitCommitGraph
+          commits={pastCommits.map(commit => ({
+            sha: commit.commit,
+            parents: commit.pre_commits
+          }))}
+          getNodeHeight={(sha: string) => nodeHeights[sha] ?? 55}
+        />
+      )}
+      <ol className={historyDialogBoxStyle}>
+        {pastCommits.map((commit, index) => {
+          return (
+            <SingleCommitNode
+              key={index}
+              commit={commit}
+              baseCommitId={baseCommitId}
+              trans={props.trans}
+              setRef={node => {
+                nodes.current[commit.commit] = node;
+              }}
+              updateBaseCommitId={updateBaseCommitId}
+            ></SingleCommitNode>
+          );
+        })}
+      </ol>
+    </div>
+  );
+};
+
+export const SingleCommitNode: React.FunctionComponent<ISingleCommitProps> = (
+  props: ISingleCommitProps
+): React.ReactElement => {
+  const { baseCommitId, updateBaseCommitId } = props;
+  const isLatest = props.commit.commit === baseCommitId;
+
+  let isBold;
+  if (isLatest) {
+    isBold = true;
+  }
+  /**
+   * Returns a callback which is invoked upon clicking a commit name.
+   *
+   * @param commit - commit
+   * @returns callback
+   */
+  const onCommitClickFactory = (commitId: string) => {
```

Comment:
[optional] it is a good practice to wrap callback in [useCallback](https://react.dev/reference/react/useCallback) hook to cache their definition between rendering.